internal/controllers/iam/user_controller.go

```diff
@@ -0,0 +1,180 @@
+type UserController struct {
+	Client     client.Client
+	Finalizers finalizer.Finalizers
+}
+
+// userFinalizer implements the finalizer.Finalizer interface for User cleanup.
+// This is used to clean up associated UserDeactivations when a User is deleted.
+type userFinalizer struct {
+	Client client.Client
+}
+
+// Finalize implements the finalizer.Finalizer interface.
+// TODO: Implement this
+func (f *userFinalizer) Finalize(ctx context.Context, obj client.Object) (finalizer.Result, error) {
+	return finalizer.Result{}, nil
+}
+
+// +kubebuilder:rbac:groups=iam.miloapis.com,resources=users,verbs=get;list;watch;update
+// +kubebuilder:rbac:groups=iam.miloapis.com,resources=users/status,verbs=update
+// +kubebuilder:rbac:groups=iam.miloapis.com,resources=userdeactivations,verbs=get;list;watch
+
+// Reconcile is the main reconciliation loop for the UserController.
+func (r *UserController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := log.FromContext(ctx).WithName("user-controller")
+	log.Info("Starting reconciliation", "request", req.Name)
+
+	user := &iamv1alpha1.User{}
+	if err := r.Client.Get(ctx, types.NamespacedName{Name: req.Name}, user); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, fmt.Errorf("failed to get User: %w", err)
+	}
+	log.Info("reconciling User", "user", user.Name)
+
+	// Handle finalizer lifecycle and ensure it is set.
+	finalizeResult, err := r.Finalizers.Finalize(ctx, user)
+	if err != nil {
+		log.Error(err, "failed to run finalizers for User")
+		return ctrl.Result{}, fmt.Errorf("failed to run finalizers for User: %w", err)
+	}
+	if finalizeResult.Updated {
+		if err := r.Client.Update(ctx, user); err != nil {
+			log.Error(err, "failed to update User after finalizer update")
+			return ctrl.Result{}, fmt.Errorf("failed to update User after finalizer update: %w", err)
+		}
+		return ctrl.Result{}, nil
+	}
+
+	// Stop reconciling if deletion in progress.
+	if !user.DeletionTimestamp.IsZero() {
+		log.Info("User is being deleted, skipping reconciliation", "user", user.Name)
+		return ctrl.Result{}, nil
+	}
+
+	// Determine desired state based on existence of any UserDeactivation for this user
+	var udList iamv1alpha1.UserDeactivationList
+	if err := r.Client.List(ctx, &udList, client.MatchingFields{"spec.userRef.name": user.Name}); err != nil {
+		log.Error(err, "failed to list UserDeactivations")
+		return ctrl.Result{}, fmt.Errorf("failed to list UserDeactivations: %w", err)
+	}
+
+	// Capture the current status to detect changes later
+	oldUserStatus := user.Status.DeepCopy()
+
+	// Defining the desired user state
+	var desiredState iamv1alpha1.UserState
+	// If there are any UserDeactivations for this user, the user is inactive
+	if len(udList.Items) > 0 {
+		desiredState = iamv1alpha1.UserStateInactive
+	} else {
+		desiredState = iamv1alpha1.UserStateActive
+	}
+	user.Status.State = desiredState
```

Comment:
We should set this based on the status of the UserDeactivation. If the UserDeactivation hasn't been processed yet, then the user will still be active.